torchmultimodal/modules/layers/transformer.py

```diff
@@ -0,0 +1,327 @@
+        hidden_states,
+        attention_mask=None,
+        head_mask=None,
+    ):
+        self_outputs = self.attention(
+            hidden_states,
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+        )
+
+        attention_output = self.dropout(self.output(self_outputs[0]))
+
+        outputs = (attention_output,) + self_outputs[
+            1:
+        ]  # add attentions if we output them
+        return outputs
+
+
+class TransformerLayer(nn.Module):
+    def __init__(
+        self,
+        hidden_size: int = 768,
+        num_attention_heads: int = 12,
+        hidden_dropout_prob: float = 0.0,
+        intermediate_size: int = 3072,
+        intermediate_activation: Callable[..., Tensor] = nn.functional.gelu,
+        attention_probs_dropout_prob: float = 0.0,
+        layer_norm_eps: float = 1e-12,
+    ):
+        super().__init__()
+        self.attention = TransformerAttention(
+            hidden_size=hidden_size,
+            num_attention_heads=num_attention_heads,
+            hidden_dropout_prob=hidden_dropout_prob,
+            attention_probs_dropout_prob=attention_probs_dropout_prob,
+        )
+        self.intermediate = nn.Linear(hidden_size, intermediate_size)
+        self.intermediate_activation = intermediate_activation
+        self.output = nn.Linear(intermediate_size, hidden_size)
+        self.dropout = nn.Dropout(hidden_dropout_prob)
+        self.layernorm_before = Fp32LayerNorm(hidden_size, eps=layer_norm_eps)
+        self.layernorm_after = Fp32LayerNorm(hidden_size, eps=layer_norm_eps)
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        head_mask=None,
+    ):
+        # TODO(asg): Support postnorm transformer architecture
+        # TODO(asg): After verification with this code, try replacing with
+        # torchtext transformer implementation
+        hs = self.layernorm_before(hidden_states)
+        self_attention_outputs = self.attention(
+            hs,  # in ViT, layernorm is applied before self-attention
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+        )
+        attention_output = self_attention_outputs[0]
+        outputs = self_attention_outputs[
+            1:
+        ]  # add self attentions if we output attention weights
+
+        # first residual connection
+        hidden_states = attention_output + hidden_states
+
+        # in ViT, layernorm is also applied after self-attention
+        layer_output = self.layernorm_after(hidden_states)
+
+        layer_output = self.intermediate(layer_output)
+        layer_output = self.intermediate_activation(layer_output)
+
+        # second residual connection is done here
+        layer_output = self.output(layer_output)
+        layer_output = self.dropout(layer_output)
+        layer_output += hidden_states
+
+        outputs = (layer_output,) + outputs
+
+        return outputs
+
+
+class TransformerEncoder(nn.Module):
```

Comment:
Just to be clear, this is a pre-norm transformer and not a generally used postnorm architecture. Depending on how torchmm plans to support them, you can make it more specific.

Reply:
@apsdehal yes we have plans on have a generic transformer implementation.
 This task is simply a first step to move out the transformer and write tests to ensure we dont break anything during later refactoring. Happy to walk you through our plans  in our 1:1